### ISSUE TITLE
Scaling the dot size based on number of points

### ISSUE THREAD

**Issue by gokceneraslan** [chanzuckerberg/cellxgene#759]
It'd be really nice to decrease the dotsize when visualizing large datasets and increase it for smaller ones.

We have a simple scaling in scanpy (https://github.com/theislab/scanpy/blob/master/scanpy/plotting/_anndata.py#L323) which is `120000 / n` where n is the number of cells. This leads to very big dots when dataset is very small, so you might want to limit it via min() or max().

**Comment by gokceneraslan** [comment]
(Discussed with @colinmegill )

**Comment by colinmegill** [comment]
@bkmartinjr @liaprins-czi this will need to take into consideration all three dot sizes: 

* default
* deselected
* hover

**Comment by colinmegill** [comment]
First thought - we might hold `deselected` size constant, at the size it is now, and scale `default` and `hover` together, with a min and a max for both TBD

**Comment by liaprins-czi** [comment]
Yeah I think that ^^^ would be good. It'd be confusing to have the deselected size scale up for a small dataset and end up the same size as a default dot of a large dataset.

**Comment by bkmartinjr** [comment]
Should the scaling be based upon number of points, canvas/window size, zoom or some combination?

Right now, scaling occurs based upon canvas/window size, but not the others (assuming that the canvas has a fixed [-1,1] coordinate space.

**Comment by liaprins-czi** [comment]
I created a Figma board to start loosely prototyping what each of the possible scaling methods would look like that @bkmartinjr mentioned above:
- number of points
- canvas/window size
- zoom level

https://www.figma.com/file/dY5jElMTaYU2XqQLBhi1Et/Scale-dot-size?node-id=0%3A1

I haven't tried combining methods yet. I just wanted to have visuals for how each method might behave on its own, to start with.

**Comment by liaprins-czi** [comment]
Update: Now I have tried out really rough conceptual mockups (in the Figma file linked to above) of "dot density" (number of dots / window size*). This way we account for both at once, which I think makes sense. We can also still apply zoom scaling (or not apply zoom scaling, see mocks of both in the Figma) on top of the dot density approach.

Assuming we wanted to use dot density as the scaling method, we'd then just need to decide:

1. Allow zooming to scale dots or just scale vis as a whole (and maintain size of dots)? There are pros and cons to either.
2. _How_ to scale by dot size and accommodate the ranges of cell count in various datasets. (120,000 / n) has limitations after about 100,000 cells (or even before, if we want to reserve 1px diameter for dots that are turned "off"). 
- Let's just say hypothetically the bottom end of the range count that we want to accommodate for is 3,000 cells, and the upper end is 1,000,000. 
- Then let's say we choose the smallest and largest window sizes we care about _optimizing_ for; let's say 720px tall and 2,000px tall just for having numbers to play with.
- And let's say we choose the smallest optimal diameter to be 2px and the largest optimal diameter to be 12px.
- We can map the highest density we want to optimize for (1,000,000 cells / a window 720px tall = **~1,388.9 dots/px height**) to the smallest dot diameter we want to optimize for (**2px**), and the lowest density (3,000 cells / 2000px = **1.5 dots/px height**) to the largest diameter (**12px**). This could be a linear mapping I think.

(Note that those numbers are just placeholders we'd have to decide about what measurements and counts we care about optimizing for, and what the smallest optimal and largest optimal dot size would be. Given that mapping range, we'd still be able to visualize dots for cell counts and window sizes beyond what we care about supporting for an optimal experience, I think....)

What do others think about this?
